File: v1-migration-backend/src/batcher.ts
import type { Env } from "./env";
import { createDb } from "./db/client";

interface QueueItem {
  photoId: string;
  storageUrl: string;
}

export class PhotoBatcher implements DurableObject {
  private queue: QueueItem[] = [];
  private alarmScheduled = false;

  constructor(
    private readonly state: DurableObjectState,
    private readonly env: Env,
  ) {}

  async fetch(request: Request): Promise<Response> {
    if (request.method !== "POST") {
      return new Response("Method not allowed", { status: 405 });
    }

    const body = (await request.json()) as QueueItem;
    await this.addPhoto(body.photoId, body.storageUrl);
    return Response.json({ ok: true });
  }

  private async addPhoto(photoId: string, storageUrl: string): Promise<void> {
    this.queue.push({ photoId, storageUrl });
    const batchSize = Number.parseInt(this.env.BATCH_SIZE || "10", 10);

    if (this.queue.length >= batchSize) {
      await this.flush();
      return;
    }

    if (!this.alarmScheduled) {
      this.alarmScheduled = true;
      const timeoutMs = Number.parseFloat(this.env.BATCH_TIMEOUT_MS || "50");
      await this.state.storage.setAlarm(Date.now() + timeoutMs);
    }
  }

  async alarm(): Promise<void> {
    this.alarmScheduled = false;
    if (this.queue.length > 0) {
      await this.flush();
    }
  }

  private async flush(): Promise<void> {
    const batchItems = [...this.queue];
    this.queue = [];
    this.alarmScheduled = false;

    if (batchItems.length === 0) return;

    const payload = {
      images: batchItems.map((item) => ({
        image_id: item.photoId,
        url: item.storageUrl,
      })),
    };

    const photoIds = batchItems.map((item) => item.photoId);
    await this.updatePhotosStatus(photoIds, "processing");

    const batchId = crypto.randomUUID();
    const runsyncUrl = `${this.env.INFERENCE_URL.replace(/\/$/, "")}/runsync`;
    const runsyncPayload = {
      input: {
        images: payload.images,
        batch_id: batchId,
        callback_url: this.env.API_CALLBACK_URL,
      },
    };

    try {
      const inferenceResponse = await fetch(runsyncUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(runsyncPayload),
      });

      if (!inferenceResponse.ok) {
        const text = await inferenceResponse.text();
        console.error(`Inference failed: ${inferenceResponse.status} ${text}`);
        await this.updatePhotosStatus(photoIds, "failed");
        return;
      }

      const jobResult = (await inferenceResponse.json()) as {
        error?: string;
        output?: {
          error?: string;
          acknowledged?: boolean;
          batch_id?: string;
          processed_images?: number;
        };
      };
      const output = jobResult.output || {};

      if (jobResult.error || output.error) {
        console.error(`Inference job failed: ${JSON.stringify(jobResult)}`);
        await this.updatePhotosStatus(photoIds, "failed");
        return;
      }

      console.log(
        `Inference batch complete (batch_id=${output.batch_id}, processed=${output.processed_images})`,
      );
    } catch (error) {
      console.error("Error in batch inference runner:", error);
      await this.updatePhotosStatus(photoIds, "failed");
    }
  }

  private async updatePhotosStatus(
    photoIds: string[],
    status: string,
  ): Promise<void> {
    if (photoIds.length === 0) return;

    const sql = createDb(this.env);
    try {
      await sql`
        UPDATE photos
        SET status = ${status}
        WHERE id IN ${sql(photoIds)}
      `;
    } catch (error) {
      console.error("Failed to update photos status:", error);
    } finally {
      await sql.end({ timeout: 5 });
    }
  }
}

export async function enqueuePhoto(
  env: Env,
  photoId: string,
  storageUrl: string,
): Promise<void> {
  const id = env.BATCHER.idFromName("global");
  const stub = env.BATCHER.get(id);
  await stub.fetch("http://batcher/add", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ photoId, storageUrl }),
  });
}

File: v1-migration-backend/src/batcher.py
"""Photo batching Durable Object + inference dispatch."""

from __future__ import annotations

import json
import uuid

from workers import DurableObject, Response, fetch

from db import create_db


class PhotoBatcher(DurableObject):
    def __init__(self, ctx, env):
        super().__init__(ctx, env)
        self.queue: list[dict[str, str]] = []
        self.alarm_scheduled = False

    async def fetch(self, request):
        if request.method != "POST":
            return Response("Method not allowed", status=405)
        body = await request.json()
        await self.add_photo(body["photoId"], body["storageUrl"])
        return Response.json({"ok": True})

    async def add_photo(self, photo_id: str, storage_url: str) -> None:
        self.queue.append({"photoId": photo_id, "storageUrl": storage_url})
        batch_size = int(getattr(self.env, "BATCH_SIZE", None) or "10")

        if len(self.queue) >= batch_size:
            await self.flush()
            return

        if not self.alarm_scheduled:
            self.alarm_scheduled = True
            timeout_ms = float(getattr(self.env, "BATCH_TIMEOUT_MS", None) or "50")
            await self.ctx.storage.setAlarm(int(__import__("time").time() * 1000) + int(timeout_ms))

    async def alarm(self) -> None:
        self.alarm_scheduled = False
        if self.queue:
            await self.flush()

    async def flush(self) -> None:
        batch_items = list(self.queue)
        self.queue = []
        self.alarm_scheduled = False
        if not batch_items:
            return

        photo_ids = [item["photoId"] for item in batch_items]
        await self.update_photos_status(photo_ids, "processing")

        batch_id = str(uuid.uuid4())
        runsync_url = f"{self.env.INFERENCE_URL.rstrip('/')}/runsync"
        payload = {
            "input": {
                "images": [
                    {"image_id": item["photoId"], "url": item["storageUrl"]}
                    for item in batch_items
                ],
                "batch_id": batch_id,
                "callback_url": self.env.API_CALLBACK_URL,
            }
        }

        try:
            response = await fetch(
                runsync_url,
                method="POST",
                headers={"Content-Type": "application/json"},
                body=json.dumps(payload),
            )
            if not response.ok:
                text = await response.text()
                print(f"Inference failed: {response.status} {text}")
                await self.update_photos_status(photo_ids, "failed")
                return

            job_result = await response.json()
            output = job_result.get("output", {})
            if job_result.get("error") or output.get("error"):
                print(f"Inference job failed: {job_result}")
                await self.update_photos_status(photo_ids, "failed")
                return

            print(
                f"Inference batch complete (batch_id={output.get('batch_id')}, "
                f"processed={output.get('processed_images')})"
            )
        except Exception as exc:
            print(f"Error in batch inference runner: {exc}")
            await self.update_photos_status(photo_ids, "failed")

    async def update_photos_status(self, photo_ids: list[str], status: str) -> None:
        if not photo_ids:
            return
        db = create_db(self.env.DATABASE_URL)
        try:
            placeholders = ", ".join(f"'{pid}'::uuid" for pid in photo_ids)
            db.execute(f"UPDATE photos SET status = %s WHERE id IN ({placeholders})", (status,))
        except Exception as exc:
            print(f"Failed to update photos status: {exc}")
        finally:
            db.close()


async def enqueue_photo(env, photo_id: str, storage_url: str) -> None:
    stub = env.BATCHER.get(env.BATCHER.idFromName("global"))
    await stub.fetch(
        "http://batcher/add",
        method="POST",
        headers={"Content-Type": "application/json"},
        body=json.dumps({"photoId": photo_id, "storageUrl": storage_url}),
    )
